R-packages/evalcast/tests/testthat/test-download_signal.R

```diff
@@ -0,0 +1,146 @@
+    )
+  })
+
+  expect_called(mock_covidcast_signal, 2)
+  expect_equal(downloaded_data, create_fake_downloaded_signal(c("al", "wy"), 10))
+})
+
+
+test_that("download_signals caches when called with cache dir", {
+  # Mock out the call to `covidcast_signal_wrapper()`.
+  mock_covidcast_signal <- mock(
+    create_fake_downloaded_signal(c("al", "wy"), 5),
+    create_fake_downloaded_signal(c("al", "wy"), 10)
+  )
+
+  temp_offline_signal_dir <- fs::dir_create(withr::local_file(tempfile()))
+
+  # Call once.
+  mockr::with_mock(covidcast_signal_wrapper = mock_covidcast_signal, {
+    downloaded_data <- download_signal(
+      data_source = "source",
+      signal = "signal",
+      geo_type = "state",
+      geo_values = c("al", "wy"),
+      end_day = as.Date("2020-01-01"),
+      issue = as.Date("2020-01-02"),
+      offline_signal_dir = temp_offline_signal_dir
+    )
+  })
+
+  expect_called(mock_covidcast_signal, 1)
+  expect_equal(downloaded_data, create_fake_downloaded_signal(c("al", "wy"), 5))
+
+  # Call a second time.
+  mockr::with_mock(covidcast_signal_wrapper = mock_covidcast_signal, {
+    downloaded_data <- download_signal(
+      data_source = "source",
+      signal = "signal",
+      geo_type = "state",
+      geo_values = c("al", "wy"),
+      end_day = as.Date("2020-01-01"),
+      issue = as.Date("2020-01-02"),
+      offline_signal_dir = temp_offline_signal_dir
+    )
+  })
+
+  expect_called(mock_covidcast_signal, 1)
+  expect_equal(downloaded_data, create_fake_downloaded_signal(c("al", "wy"), 5))
+})
+
+
+test_that("populate_cache calls mock_covidcast_signal appropriately", {
+  # Mock out the call to `covidcast_signal_wrapper()`.
+  mock_covidcast_signal <- mock(
+    create_fake_downloaded_signal(c("al", "wy"), 5),
+    create_fake_downloaded_signal(c("al", "wy"), 10)
+  )
+
+  temp_offline_signal_dir <- fs::dir_create(withr::local_file(tempfile()))
+
+  # Call once.
+  mockr::with_mock(covidcast_signal_wrapper = mock_covidcast_signal, {
+    populate_cache(tibble(
+      data_source = c("hhs", "jhu-csse"),
+      signal = c("confirmed_admissions_covid_1d", "confirmed_incidence_num"),
+      start_day = c("2020-08-01", "2020-08-01"),
+      as_of = c("2022-01-01", "2022-01-01"),
+      geo_type = c("state", "state")
+    ), offline_signal_dir = temp_offline_signal_dir)
+  })
+
+  expect_called(mock_covidcast_signal, 2)
+  expect_equal(mock_args(mock_covidcast_signal), list(
+      list(
+        data_source = "hhs",
+        signal = "confirmed_admissions_covid_1d",
+        start_day = "2020-08-01",
+        end_day = "2022-01-01",
+        geo_type = "state",
+        geo_values = "*",
+        as_of = "2022-01-01"
+      ),
+      list(
+        data_source = "jhu-csse",
+        signal = "confirmed_incidence_num",
+        start_day = "2020-08-01",
+        end_day = "2022-01-01",
+        geo_type = "state",
+        geo_values = "*",
+        as_of = "2022-01-01"
+      )
+    )
+  )
+})
```

Comment:
Nice tests!  Running them + others and looking at CI yields some more annoying cleanup to do:
- "Unknown or uninitialised column:" many times: resolved by moving away from `is.null(tbl$maybecol)`
- Need to expect warning at
  - at test-download_signal.R:115:4 ("cache ends earlier than")
  - at test-predictions.R:51:4 ("only selects weekly (Saturday) target end dates")
  - at test-predictions.R:154:4 (same)
  - at test-predictions.R:210:4 (same)
  - at test-predictions.R:291:4 (same)
- test failure: need to add time_type = "day" to expected mock args in populate_cache test.
  
Other CI checks I am able to understand / guess at:
```
Warning: Undocumented arguments in documentation object 'download_signal'
  ‘time_type’
```
^ Self-explanatory

```
Warning: get_predictions.Rd:117: Bad \link text
problem found in ‘get_predictions.Rd’
```
^ Looks like `\code` needs to be outside `\link` when linking to bettermc::mclapply for `\link` to work.  If we have `@md` on this entry or roxygen markdown mode enabled for the package, then it could just be
```
[`bettermc::mclapply`]
```
instead.